### PR TITLE
2015 10 mn edit button

### PR DESCRIPTION
**Pull request by MagdaN**
OK, I added a simple directive that will show an edit and an delete button when the user is admin or creator of a resource. I assume that the buttons are added on the detail page of a resource and that the creator is known (means in scope) of that page. You can add the buttons like that:

```
<edit-buttons 
data-userpath="{{ data.creator_path }}" 
data-editurl="#!/metrics/{{data.id}}/edit"                
data-deletefunction="delete">
</edit-buttons>
```

*  data-userpath is the creator of the resource (should be a path),
*  data-editurl is the url where the edit button should link to
*  deletefunction is the function from the surrounding scope that is called when clicking the 'delete' button

### CODE REVIEW

**Comment by FabiApfelkern** [comment]
@MagdaN There might be a bug in it. Instead of Auth.userpath === scope.userpath; it should be
Auth.state.userPath === scope.userpath; ?

**Comment by MagdaN** [comment]
@FabiApfelkern : yes, thank!
@nidico and @slomo : it seems that the adhocracy api has changed - instead of the userpath the full userurl is sent. Why is this?

**Comment by nidico** [comment]
> @nidico and @slomo : it seems that the adhocracy api has changed - instead of the userpath the full userurl is sent. Why is this?

This has always been the case, at least since 2014.

**Comment by slomo** [comment]
Nope. I strongly disagree.

**Comment by cbotsikas** [comment]
so is it safe to use RegexValidator("^(/[^/]*)+/?$") for the userpath at the django models? eg https://github.com/policycompass/policycompass-services/blob/master/apps/indicatorservice/models.py#L22 https://github.com/policycompass/policycompass-services/blob/master/apps/metricsmanager/models.py#L9

**Comment by FabiApfelkern** [comment]
@cbotsikas I don't understand your question.

**Comment by cbotsikas** [comment]
@FabiApfelkern I'm confused with @MagdaN statement "adhocracy api has changed - instead of the userpath the full userurl is sent". My question is, what's expected as userpath in the django backend when self.request.user.resource_path is used? Is is a userpath in the form of "/principals/users/XXXXXX" or a full url?

**Comment by FabiApfelkern** [comment]
Good question. I just copied the RegexValidator But as the full path comes with a request, we are all storing this full path at the moment. @nidico This was the plan, wasn't it?

**Comment by MagdaN** [comment]
it obviously changed - take a look here at the staging server under /api/v1/metricsmanager/metrics

**Comment by cbotsikas** [comment]
@MagdaN You are right... here: https://services-stage.policycompass.eu/api/v1/metricsmanager/metrics I can see both "creator_path": "/principals/users/000001" and "creator_path": "https://adhocracy-frontend-stage.policycompass.eu/api/principals/users/0000001/". 
So I guess the regular expression ^(/[^/]*)+/?$ is invalid. 
Shall we remove it from models and keep the CharField without validation or change the regex? @nidico @slomo @FabiApfelkern any thoughts?

**Comment by FabiApfelkern** [comment]
I have to admit I have problems with regexes one the first sight. So I just ask: What is the exact intention of that one? It definitely allows to store the complete path.

**Comment by cbotsikas** [comment]
@FabiApfelkern It's allowed because we don't check for validity.. According to [Django's documentation](https://docs.djangoproject.com/en/1.8/ref/validators/#how-validators-are-run), Model validators run automatically when used with ModelForm. In our case with REST framwork [serializers should handle this](http://www.django-rest-framework.org/api-guide/validators/) but I think [based on this post](https://groups.google.com/forum/#!topic/django-rest-framework/siPqtlp6DpA) we need to manually call   and handle serializer.is_valid() inside the pre_save() method since it's automatically checked before calling pre_save(). 
In conclusion, i think that since we set the value of the user_path inside pre_save(), validation is skipped so we should handle it there manually...

**Comment by cbotsikas** [comment]
@FabiApfelkern As for regular expressions in general, I find them really powerful and helpful. It's not that hard once you grasp the logic. But simply reading the syntax wont help much. I strongly recommend you to check [Expresso](http://www.ultrapico.com/expresso.htm), [Regexper](http://regexper.com/), [regexpal](http://regexpal.com/) and [pythex](http://pythex.org/) Note that there are some syntactic differences between languages so in our case  "^(/[^/]*)+/?$" is working with pythex but not with Regexper where you need escape slash like this: "^(\/[^\/]*)+\/?$" ([direct link](http://regexper.com/#%5E(%5C%2F%5B%5E%5C%2F%5D*)%2B%5C%2F%3F%24))

**Comment by cbotsikas** [comment]
btw, @FabiApfelkern here is the code executed when using the (List)CreateAPIView https://github.com/tomchristie/django-rest-framework/blob/version-2.4.x/rest_framework/mixins.py#L49

**Comment by FabiApfelkern** [comment]
@cbotsikas Thank you for the detailed information. I also think that regexes are powerful I had them a lot in uni. What I was saying is that I sometimes cannot understand them by just reading them, However, we should decide how to proceed. Are we storing the absolute path, the relative path or even just the ID? The later two would make it a little bit more environment independent.

**Comment by MagdaN** [comment]
@FabiApfelkern @cbotsikas  so you would suggest to use just the path? This would be a little bit more complicated as A3 sends the full url. We would have to get the path from the url in the presave method and in the directive from this pull request (and probably other directives) we have to adjust  how we check if the current user is also the owner of a resource.

**Comment by FabiApfelkern** [comment]
I would say we keep everything as it is for now. But lets discuss later again in the telco. I merge this for now.
